# **1. Import Library**

Pada tahap ini, Anda perlu mengimpor beberapa pustaka (library) Python yang dibutuhkan untuk analisis data dan pembangunan model machine learning.

In [1]:
import pandas as pd
import numpy as np
import warnings
import joblib
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, f1_score, precision_score, recall_score, classification_report, confusion_matrix
from sklearn.linear_model import Perceptron, RidgeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis
from sklearn.model_selection import GridSearchCV, cross_val_score

warnings.filterwarnings('ignore')

# **2. Memuat Dataset dari Hasil Clustering**

Memuat dataset hasil clustering dari file CSV ke dalam variabel DataFrame.

In [2]:
data = pd.read_csv("Adult Income Dataset/final_data.csv")
data

education_num  capital_gain  hours_per_week  sex  age  cluster
0                 13   1092.841202              40    1   39        3
1                 13      0.000000              13    1   50        0
2                  9      0.000000              40    1   38        0
3                  7      0.000000              40    1   53        0
4                 13      0.000000              40    0   28        1
...              ...           ...             ...  ...  ...      ...
45189             13      0.000000              40    1   33        0
45190             13      0.000000              36    0   39        1
45191             13      0.000000              50    1   38        0
45192             13   1120.673663              40    1   44        3
45193             13      0.000000              60    1   35        0

[45194 rows x 6 columns]

# **3. Data Splitting**

Tahap Data Splitting bertujuan untuk memisahkan dataset menjadi dua bagian: data latih (training set) dan data uji (test set).

In [3]:
X = data.drop('cluster', axis=1)
y = data['cluster']

In [4]:
X

education_num  capital_gain  hours_per_week  sex  age
0                 13   1092.841202              40    1   39
1                 13      0.000000              13    1   50
2                  9      0.000000              40    1   38
3                  7      0.000000              40    1   53
4                 13      0.000000              40    0   28
...              ...           ...             ...  ...  ...
45189             13      0.000000              40    1   33
45190             13      0.000000              36    0   39
45191             13      0.000000              50    1   38
45192             13   1120.673663              40    1   44
45193             13      0.000000              60    1   35

[45194 rows x 5 columns]

In [5]:
y

0        3
1        0
2        0
3        0
4        1
        ..
45189    0
45190    1
45191    0
45192    3
45193    0
Name: cluster, Length: 45194, dtype: int64

In [6]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# **4. Membangun Model Klasifikasi**


## **a. Membangun Model Klasifikasi**

Setelah memilih algoritma klasifikasi yang sesuai, langkah selanjutnya adalah melatih model menggunakan data latih.

Berikut adalah rekomendasi tahapannya.
1. Pilih algoritma klasifikasi yang sesuai, seperti Logistic Regression, Decision Tree, Random Forest, atau K-Nearest Neighbors (KNN).
2. Latih model menggunakan data latih.

In [15]:
models = {
    "K-Nearest Neighbors": KNeighborsClassifier(),
    "Ridge Classifier": RidgeClassifier(),
    "Perceptron": Perceptron(),
}

In [16]:
trained_models = {}
for model_name, model in models.items():
    model.fit(X_train, y_train)
    trained_models[model_name] = model

In [9]:
for model_name, model in trained_models.items():
    joblib.dump(model, f'Original Model/{model_name}_trained.pkl')

## Penjelasan Algoritma

Dalam proyek ini, kami menggunakan beberapa algoritma machine learning untuk melakukan klasifikasi pada dataset yang terdiri dari 4 kelas. Algoritma yang diterapkan mencakup K-Nearest Neighbors (KNN), Ridge Classifier, dan Perceptron. 

1. **K-Nearest Neighbors (KNN)**: KNN adalah algoritma non-parametrik yang mengklasifikasikan data berdasarkan kedekatan dengan data lain. Algoritma ini menghitung jarak antara titik data baru dan titik data dalam dataset pelatihan, kemudian memilih `k` tetangga terdekat untuk menentukan kelas berdasarkan mayoritas suara.

2. **Ridge Classifier**: Algoritma ini merupakan varian dari regresi linier yang mengandung penalti L2 untuk mengurangi kompleksitas model dan mencegah overfitting. Ridge Classifier sangat cocok untuk klasifikasi dengan fitur yang berkorelasi tinggi.

3. **Perceptron**: Perceptron adalah algoritma pembelajaran supervised yang digunakan untuk klasifikasi biner, tetapi dapat diperluas untuk klasifikasi multi-kelas. Ia bekerja dengan mengupdate bobot berdasarkan kesalahan klasifikasi hingga konvergensi tercapai. 

Dataset ini terdiri dari 4 kelas, dan semua algoritma ini telah dioptimalkan untuk mencapai kinerja klasifikasi yang lebih baik. Evaluasi dilakukan menggunakan metrik seperti akurasi, precision, recall, dan weighted F1 score untuk mendapatkan gambaran yang jelas mengenai performa masing-masing algoritma.

## **b. Evaluasi Model Klasifikasi**

Berikut adalah **rekomendasi** tahapannya.
1. Lakukan prediksi menggunakan data uji.
2. Hitung metrik evaluasi seperti Accuracy dan F1-Score (Opsional: Precision dan Recall).
3. Buat confusion matrix untuk melihat detail prediksi benar dan salah.

In [10]:
for model_name, model in trained_models.items():
    y_pred = model.predict(X_test)

    train_accuracy = accuracy_score(y_train, model.predict(X_train))
    test_accuracy = accuracy_score(y_test, y_pred)

    f1 = f1_score(y_test, y_pred, average='weighted')
    precision = precision_score(y_test, y_pred, average='weighted')
    recall = recall_score(y_test, y_pred, average='weighted')

    conf_matrix = confusion_matrix(y_test, y_pred)

    total_accuracy = (train_accuracy + test_accuracy) / 2

    print(f'{model_name} Total Accuracy: {total_accuracy:.4f}')
    print(f'{model_name} Weighted F1 Score: {f1:.4f}')
    print(f'{model_name} Precision: {precision:.4f}')
    print(f'{model_name} Recall: {recall:.4f}')
    print(f'{model_name} Confusion Matrix:\n{conf_matrix}')
    print(f'{model_name} Classification Report:')
    print(classification_report(y_test, y_pred))
    print('-' * 60)

K-Nearest Neighbors Total Accuracy: 0.9455
K-Nearest Neighbors Weighted F1 Score: 0.9284
K-Nearest Neighbors Precision: 0.9296
K-Nearest Neighbors Recall: 0.9304
K-Nearest Neighbors Confusion Matrix:
[[5323  128    0    0]
 [ 360 2456    0    0]
 [   0    0   79  107]
 [   0    0   34  552]]
K-Nearest Neighbors Classification Report:
              precision    recall  f1-score   support

           0       0.94      0.98      0.96      5451
           1       0.95      0.87      0.91      2816
           2       0.70      0.42      0.53       186
           3       0.84      0.94      0.89       586

    accuracy                           0.93      9039
   macro avg       0.86      0.80      0.82      9039
weighted avg       0.93      0.93      0.93      9039

------------------------------------------------------------
Ridge Classifier Total Accuracy: 0.9805
Ridge Classifier Weighted F1 Score: 0.9695
Ridge Classifier Precision: 0.9601
Ridge Classifier Recall: 0.9794
Ridge Classifier C

## **c. Tuning Model Klasifikasi (Optional)**

Gunakan GridSearchCV, RandomizedSearchCV, atau metode lainnya untuk mencari kombinasi hyperparameter terbaik

In [11]:
param_grids = {
    "K-Nearest Neighbors": {
        'n_neighbors': [3, 5, 7, 10],
        'weights': ['uniform', 'distance'],
        'algorithm': ['auto', 'ball_tree', 'kd_tree']
    },
    "Ridge Classifier": {
        'alpha': [0.01, 0.1, 1.0, 10.0, 100.0]
    },
    "Perceptron": {
        'max_iter': [1000, 2000, 3000],
        'tol': [1e-3, 1e-4, 1e-5],
        'eta0': [0.01, 0.1, 0.5]
    },
}

In [12]:
best_models = {}
for model_name, model in models.items():
    grid_search = GridSearchCV(
        model, 
        param_grids[model_name], 
        cv=5, 
        scoring=['accuracy', 'f1_weighted'], 
        refit='accuracy',
        return_train_score=True
    )
    grid_search.fit(X_train, y_train)

    best_models[model_name] = grid_search.best_estimator_

    cv_scores = cross_val_score(best_models[model_name], X_train, y_train, cv=5, scoring='f1_weighted')
    print(f'{model_name} Best Parameters: {grid_search.best_params_}')
    print(f'{model_name} Cross-Validation Accuracy Score: {grid_search.cv_results_["mean_test_accuracy"].mean():.4f}')
    print(f'{model_name} Cross-Validation F1 Score: {cv_scores.mean():.4f}')
    print('-' * 60)

    joblib.dump(best_models[model_name], f'Tuning Model/{model_name}_best.pkl')

K-Nearest Neighbors Best Parameters: {'algorithm': 'ball_tree', 'n_neighbors': 3, 'weights': 'distance'}
K-Nearest Neighbors Cross-Validation Accuracy Score: 0.9313
K-Nearest Neighbors Cross-Validation F1 Score: 0.9409
------------------------------------------------------------
Ridge Classifier Best Parameters: {'alpha': 0.01}
Ridge Classifier Cross-Validation Accuracy Score: 0.9816
Ridge Classifier Cross-Validation F1 Score: 0.9727
------------------------------------------------------------
Perceptron Best Parameters: {'eta0': 0.1, 'max_iter': 1000, 'tol': 0.001}
Perceptron Cross-Validation Accuracy Score: 0.9780
Perceptron Cross-Validation F1 Score: 0.9737
------------------------------------------------------------


## **d. Evaluasi Model Klasifikasi setelah Tuning (Optional)**

Berikut adalah rekomendasi tahapannya.
1. Gunakan model dengan hyperparameter terbaik.
2. Hitung ulang metrik evaluasi untuk melihat apakah ada peningkatan performa.

In [13]:
for model_name, model in best_models.items():
    y_pred = model.predict(X_test)

    total_accuracy = accuracy_score(y_test, y_pred)

    f1 = f1_score(y_test, y_pred, average='weighted')
    precision = precision_score(y_test, y_pred, average='weighted')
    recall = recall_score(y_test, y_pred, average='weighted')

    conf_matrix = confusion_matrix(y_test, y_pred)

    print(f'{model_name} Total Accuracy: {total_accuracy:.4f}')
    print(f'{model_name} Weighted F1 Score: {f1:.4f}')
    print(f'{model_name} Precision: {precision:.4f}')
    print(f'{model_name} Recall: {recall:.4f}')
    print(f'{model_name} Confusion Matrix:\n{conf_matrix}')
    print(f'{model_name} Classification Report:')
    print(classification_report(y_test, y_pred))
    print('-' * 60)

K-Nearest Neighbors Total Accuracy: 0.9476
K-Nearest Neighbors Weighted F1 Score: 0.9464
K-Nearest Neighbors Precision: 0.9470
K-Nearest Neighbors Recall: 0.9476
K-Nearest Neighbors Confusion Matrix:
[[5339  112    0    0]
 [ 244 2572    0    0]
 [   0    0   98   88]
 [   0    0   30  556]]
K-Nearest Neighbors Classification Report:
              precision    recall  f1-score   support

           0       0.96      0.98      0.97      5451
           1       0.96      0.91      0.94      2816
           2       0.77      0.53      0.62       186
           3       0.86      0.95      0.90       586

    accuracy                           0.95      9039
   macro avg       0.89      0.84      0.86      9039
weighted avg       0.95      0.95      0.95      9039

------------------------------------------------------------
Ridge Classifier Total Accuracy: 0.9794
Ridge Classifier Weighted F1 Score: 0.9695
Ridge Classifier Precision: 0.9601
Ridge Classifier Recall: 0.9794
Ridge Classifier C

## **e. Analisis Hasil Evaluasi Model Klasifikasi**

Berikut adalah **rekomendasi** tahapannya.
1. Bandingkan hasil evaluasi sebelum dan setelah tuning (jika dilakukan).
2. Identifikasi kelemahan model, seperti:
  - Precision atau Recall rendah untuk kelas tertentu.
  - Apakah model mengalami overfitting atau underfitting?
3. Berikan rekomendasi tindakan lanjutan, seperti mengumpulkan data tambahan atau mencoba algoritma lain jika hasil belum memuaskan.

#### 1. Bandingkan hasil evaluasi sebelum dan setelah tuning

| Metric                        | K-Nearest Neighbors (Before Tuning) | K-Nearest Neighbors (After Tuning) | Ridge Classifier (Before Tuning)    | Ridge Classifier (After Tuning)     | Perceptron (Before Tuning)          | Perceptron (After Tuning)           |
|-------------------------------|--------------------------------------|-------------------------------------|--------------------------------------|-------------------------------------|--------------------------------------|-------------------------------------|
| Total Accuracy                | 0.9455                               | 0.9476                              | 0.9805                               | 0.9794                              | 0.9797                               | 0.9785                              |
| Weighted F1 Score             | 0.9284                               | 0.9464                              | 0.9695                               | 0.9695                              | 0.9698                               | 0.9698                              |
| Precision                     | 0.9296                               | 0.9470                              | 0.9601                               | 0.9601                              | 0.9633                               | 0.9633                              |
| Recall                        | 0.9304                               | 0.9476                              | 0.9794                               | 0.9794                              | 0.9785                               | 0.9785                              |

#### 2. Identifikasi kelemahan model

- **K-Nearest Neighbors**:
  - Precision dan Recall rendah untuk kelas 2.
  - Model tampaknya tidak mengalami overfitting yang signifikan, tetapi terdapat penurunan performa setelah tuning.

- **Ridge Classifier**:
  - Kelas 2 memiliki Precision dan Recall yang sangat rendah (0.00).
  - Model tetap konsisten dengan performa tinggi di kelas lainnya.

- **Perceptron**:
  - Kelas 2 juga memiliki Precision dan Recall yang sangat rendah (0.00).
  - Model menunjukkan konsistensi di kelas lainnya, tetapi ada kemungkinan underfitting untuk kelas 2.

#### 3. Rekomendasi tindakan lanjutan

- Kumpulkan data tambahan untuk kelas yang kurang terwakili, terutama untuk kelas 2.
- Pertimbangkan untuk mencoba algoritma lain atau teknik resampling seperti SMOTE untuk menangani kelas yang tidak seimbang.
- Lakukan eksplorasi lebih lanjut terhadap fitur yang digunakan untuk memahami mengapa model tidak berkinerja baik pada kelas tertentu.